samples: nrf54h20: Add sample that uses the UICR CUSTOMER infrastructure (#29145)
nrfconnect/sdk-nrf: samples/ironside_se/customer_uicr/README.rst

.. _customer_uicr_sample:

> CUSTOMER data in UICR

Reviewer: ```suggestion
CUSTOMER register data in UICR
```
Does this sample have an official "product" name?
If not, I'd edit the name as suggested.

##########################################

.. contents::
   :local:
   :depth: 2

This sample demonstrates how to store persistent data in UICR.CUSTOMER.

Requirements
************

The sample supports the following development kit:

.. table-from-sample-yaml::

Overview
********

> The sample will read CUSTOMER from UICR. During programming customer.bin is programmed to the nRF,
> which will be generated as part of uicr.hex by the gen_uicr image and flashed to the device when
> `west flash` is called.

Reviewer: ```suggestion
The sample reads the CUSTOMER register from the User information configuration registers (UICRs).
During programming, the ``gen_uicr`` image generates :file:`uicr.hex`, which includes :file:`customer.bin`.
When you run ``west flash``, :file:`uicr.hex` is programmed to the device.
```

Building and running
*********************

.. |sample path| replace:: :file:`samples/ironside_se/customer_uicr`

.. include:: /includes/build_and_run.txt

Testing
*******

After programming the sample to your development kit, complete the following steps to test it:

1. |connect_terminal|
#. Reset the development kit.

> The application writes out all the content of UICR->CUSTOMER. The written ``customer.bin`` file contains
> 0xdeadbeef as the first 4 bytes, which can be verified from the logs. The ``customer.bin`` file was
> generated using the command:

Reviewer: ```suggestion
The application writes out all the content of UICR.CUSTOMER.
The written :file:`customer.bin` file contains ``0xdeadbeef`` as the first 4 bytes, which can be verified from the logs.
To generate :file:`customer.bin`, use this command:
```

   .. code-block:: console

   python3 -c "open('customer.bin','wb').write((0xdeadbeef).to_bytes(4, 'little'))"

Configuration
*************

The sample uses the following key configurations:

Kconfig Configuration

>   The ``sysbuild/uicr/prj.conf`` file overwrites uicr image configuration. It enables CUSTOMER hex generation via CONFIG_GEN_UICR_CUSTOMER and points to the binary file
>   with CONFIG_GEN_UICR_CUSTOMER_BIN_FILE.

Reviewer: ```suggestion
  The :file:`sysbuild/uicr/prj.conf` file overrides the UICR image configuration.
  It enables CUSTOMER HEX generation with :kconfig:option:`CONFIG_GEN_UICR_CUSTOMER` and specifies the binary file with :kconfig:option:`CONFIG_GEN_UICR_CUSTOMER_BIN_FILE`.
```

Sysbuild Configuration

>   The ``sysbuild/uicr`` folder overlays content in the uicr image folder. By storing ``customer.bin`` to this folder, CONFIG_GEN_UICR_CUSTOMER_BIN_FILE can use relative
>   path to point to ``customer.bin``.

Reviewer: ```suggestion
  The :file:`sysbuild/uicr` folder overlays the content of the UICR image folder.
By storing :file:`customer.bin` in this folder, :kconfig:option:`CONFIG_GEN_UICR_CUSTOMER_BIN_FILE` can use a relative path to point to :file:`customer.bin`.
```
